src/lib.rs

```diff
@@ -1,121 +1,76 @@
-use std::{
-    fs,
-    fs::File,
-    io,
-    path::{Path, PathBuf},
-};
+use std::{fs::{File}, io, path::{Path, PathBuf}};
+use async_std::{fs::{create_dir_all, set_permissions, Permissions}};
+use zip::read::ZipFile;
 
-pub fn unzip_file(file: File, directory: &Path) {
-    let mut archive = zip::ZipArchive::new(file).unwrap();
+pub async fn unzip_archive(file: File, directory: &Path) -> Result<(), Box<dyn std::error::Error>> {
+    let mut archive = zip::ZipArchive::new(file)?;
+    let mut files = vec![];
 
     for file_index in 0..archive.len() {
-        let file = archive.by_index(file_index).unwrap();
-        unzip_archive(file, directory)
+        match archive.by_index(file_index) {
+            Ok(file) => files.push(unzip_file(file, directory).await),
```

Comment:
it does partially solves the problem with mutability and the `by_index`, we can always look into how to improve it later.